### ISSUE TITLE
`<symbol>` elements should have `aria-labelledby` attribute

### ISSUE THREAD

**Issue by jkphl** [jkphl/svg-sprite#107]
As brought up by @davidhund on [Twitter](https://twitter.com/valuedstandards/status/644125581295136768), `<symbol>` elements inside sprites could / should have an `aria-labelledby` attribute when having also a title. See [here](http://codepen.io/NathanPJF/full/GJObGm) for some further info.

**Comment by davidhund** [comment]
Awesome, do note that the linked CodePen only deals with the OSX "VoiceOver" AT. 

For more information on support I've found: http://www.paciellogroup.com/blog/2013/12/using-aria-enhance-svg-accessibility/ Especially the sections “Browser/Screen reader support for ARIA enhanced SVG” seem relevant.

Also https://github.com/jonathantneal/svg4everybody#readability-and-accessibility. Jonathan advocates the use of a `title` attribute on the `<svg/>`.

It might be good to have a small ‘Usage’ section describing two ways of implementing the svg sprites: one with `role="img"` and one with `role="presentation"`?

**Comment by jkphl** [comment]
Thanks for the resources, @davidhund! I need to read through them before I know what I've got to do ;) (Will need some time for that anyway, as I'm quite overloaded at the moment ...)

**Comment by fvsch** [comment]
It would be great to have results for other screen readers. In [my own tests](https://dl.dropboxusercontent.com/u/145744/accessible-svg-icon/test.html) I also found that VoiceOver only reads SVG icons when they're inside interactive content (`<a>` and `<button>` in my tests, there might be others too).

It's good to know that you can persuade VoiceOver to read the icon in all contexts using `aria-labelledby`. But unless other screen readers have the same issue and would benefit from the same fix, I would advise against using this fix systematically.

Additionally, there are many downsides when relying on `<title>` for accessible SVG icons:
* The title often comes from the filename, which may not be relevant
* Or the title is in a sidecar JSON file that never gets updated
* In any case, the title generally describes the graphic, and not its meaning in context
* Finally, putting the labels in the sprite file doesn't work for internationalization

**Comment by davidhund** [comment]
Good points @fvsch — Note that I have asked others on Twitter also about what they consider the most accessible markup for SVG Sprite Icons: 

I've added my current markup to a gist: https://gist.github.com/davidhund/564331193e1085208d7e
and asked (a11y)  people to respond: https://twitter.com/valuedstandards/status/647338546831982592

So far little response :(

re `<title>`: we actually have 2 places we edit the title child: in the `<symbol>` in the sprite and as child of the `svg` in which we `<use>`  our symbol. I agree that the title element in our *sprite* will often be too generic to be useful, but since we can add a title element in our `svg` elements—which we author manually most ofent—we can assume these can be more meaningful (for context), right?

That is: only if our `svg > title` is actually picked up.

For ‘presentational’ icons (most?) all this is not necessary and we can suffice with a simple `role="presentation"`.

**Comment by jkphl** [comment]
Hey @fvsch,

thanks for your input. I agree, it would be nice to have some more insights with other screen readers as well.

Regarding the `aria-labelledby` attribute, however, I already implemented this some days ago, and the attribute is available since [commit 1a231c02a0](https://github.com/jkphl/svg-sprite/commit/1a231c02a04c9cef0daf8dba68dd014f6e43749f). In fact it was not much of a change — the attribute was already implemented for all other relevant types of sprites, it only got stripped out for `<symbol>` sprites by accident.

In general, I don't think it should be up to *svg-sprite* to decide whether using a `<title>` is a good idea or not. I totally agree with @davidhund, however, that as soon as there's a `<title>` element, there should also be an `aria-labelledby` attribute (this is how it was supposed to work anyway, see above). I can't imagine any situation where this attribute would harm anyone.

Cheers,
Joschi

**Comment by fvsch** [comment]
> I agree that the title element in our sprite will often be too generic to be useful, but since we can add a title element in our svg elements—which we author manually most often—we can assume these can be more meaningful (for context), right?

In my tests if you have a `<title>` in the symbol and `<button aria-label="Relevant label"><svg><use/></svg></button>`, VoiceOver reads the aria-label and not the symbol's `<title>`. If other screen readers do too, then we can consider the `<title>` as fallback, and advise authors to use `aria-label` on their icon or the parent element. (Or to use the `title` attribute it is has wider and more consistent support.)

> I can't imagine any situation where this attribute would harm anyone.

Main thing to check would be this:

```html
<symbol id="my-icon" viewBox="0 0 16 16" aria-labelledby="my-icon-title">
    <title id="my-icon-title">Fallback label</title>
    <path />
</symbol>

<button aria-label="Preferred label">
    <svg><use xlink:href="#my-icon"></use></svg>
</button>
```

As long as screen readers don't give the priority to the fallback title because of `aria-labelledby`, we're in the clear.

**Comment by jkphl** [comment]
@fvsch  :+1: Sounds reasonable. I'm currently not able to test any screen readers though, so it'd be nice if anyone else could shed some light on this.

**Comment by fvsch** [comment]
If you look at the ARIA spec here:
http://www.w3.org/TR/wai-aria/roles#textalternativecomputation
http://www.w3.org/TR/wai-aria/states_and_properties#aria-labelledby
http://www.w3.org/TR/wai-aria/roles#img

When looking for accessible text, AT should look in this order:
- `aria-labelledby`
- `arial-label`
- the contents of the element

In the HTML [above](https://github.com/jkphl/svg-sprite/issues/107#issuecomment-143762195), when trying to read the button, the `aria-label` attribute should have precedence over the used symbol, which appears to be the case in my previous tests with VoiceOver (testing with `<title>` in the symbol). So I guess we're in the clear if the AT's ARIA implementation is correct. It would still be useful to test though.

**Comment by fvsch** [comment]
So I wrote a bunch of tests and tested in VoiceOver and JAWS, and it tends to confirm that `<title>` in `<symbol>` has little value and that the `aria-labelledby` trick on top of that is not particularly useful (I’d say it’s just a quirck in Safari/VoiceOver).
https://dl.dropboxusercontent.com/u/145744/accessible-svg-icon/test.html

Note that the `aria-labelledby` trick doesn’t hurt either, so it’s probably okay to keep it.

**Comment by jkphl** [comment]
@fvsch Thanks a lot for your efforts — this is truly valuable! :+1:

**Comment by davidhund** [comment]
@fvsch @jkphl If `<title>` and `<desc>` in `<symbol>` have limited value, why not take them out and reduce the file-size of our sprite?

And also advocate the use of `aria-label` and `role="img"`, obviously…

**Comment by jkphl** [comment]
@davidhund As far as *svg-sprite* is concerned, I always tried to keep out functionality that deals with SVG code manipulation itself. There are libraries out there that are far better with this. *svg-sprite* should focus on its aim to create sprites. Introducing the possibility to inject titles and descriptions was a kind of exception to this, as I wanted to promote using accessibility features and make it as easy as possible. I would, however, refrain from providing means for stripping titles and descriptions that are already present in the source files, at least by default. In the end one could always use one of the [transformation](https://github.com/jkphl/svg-sprite/blob/master/docs/configuration.md#shape-transformations) [hooks](https://github.com/jkphl/svg-sprite/blob/master/docs/configuration.md#svg-sprite-customization) to customize the SVG output.

As a result of all this, I will happily link to this thread as well as your and @fvsch's findings in the docs soon. I don't see any reason for further changing *svg-sprite*'s functionality at the moment though. I hope you're fine with that‽

**Comment by davidhund** [comment]
@jkphl Sure, I understand. However: you offer the option of injecting titles and descriptions that—as @fvsch has now figured out—has little value.

I completely understand you don't want to change the *svg-sprite* functionality further, but I can see some value in *removing* this functionality since it does not seem to offert the a11y benefits we thought it had…

**Comment by davidhund** [comment]
And I would like to apologise for advocating for functionality that—in hindsight—does not have the benefits I expected ;-)

**Comment by jkphl** [comment]
@davidhund No problem — I think we (i.e. you both) did some quite valuable findings, so this was worth it. ;)

Well, I get your point. However, from my perspective, even *little* value is more than *no* value, so I'd like to keep this. As long as the Paciello Group and [others](http://www.sitepoint.com/tips-accessible-svg/) promote this kind of enhancement, people will keep asking for it.

By the way I wouldn't be surprised if you (besides me) are the first *svg-sprite* user to actively use the meta data injection (I'm sure it's used by less than 1% ...). Unfortunately it's quite a geeky thing, right? And still, users don't have to use it, so there shouldn't be any harm to anyone ...

**Comment by davidhund** [comment]
:+1:
